### PR TITLE
Refactor AdfBidder

### PR DESCRIPTION
**Pull request by CTMBNara**
Rid off the OpenrtbBidder usage.

### CODE REVIEW

File: src/main/java/com/iab/openrtb/request/BidRequest.java [hunk 1]
```diff
@@ -32,6 +33,7 @@
      * Array of Imp objects (Section 3.2.4) representing the impressions
      * offered. At least 1 Imp object is required.
      */
+    @Wither
```
Comment (snahornyi) [inline src/main/java/com/iab/openrtb/request/BidRequest.java:36]: Pls remove it. It is not good to have two approaches for one action. For this one second approach is for specific case

File: src/main/java/com/iab/openrtb/request/Imp.java [hunk 2]
```diff
@@ -90,6 +91,7 @@
      * the auction. This can be useful for debugging of any issues, or for
      * optimization by the buyer.
      */
+    @Wither
```
Comment (snahornyi) [inline src/main/java/com/iab/openrtb/request/Imp.java:94]: Pls remove

File: src/main/java/org/prebid/server/bidder/adf/AdfBidder.java [hunk 3]
```diff
@@ -1,56 +1,162 @@
...
+            new TypeReference<>() {
+            };
 
-    public AdfBidder(String endpointUrl, JacksonMapper mapper) {
-        super(endpointUrl, RequestCreationStrategy.SINGLE_REQUEST, ExtImpAdf.class, mapper);
+    private final String endpointUrl;
+    private final JacksonMapper mapper;
+
+    public AdfBidder(@NonNull String endpointUrl,
+                     @NonNull JacksonMapper mapper) {
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/bidder/adf/AdfBidder.java:46]: Pls use project style with requireNonNull function

File: src/main/java/org/prebid/server/bidder/adf/AdfBidder.java [hunk 4]
```diff
@@ -1,56 +1,162 @@
...
+                .map(imp -> modifyImp(imp, errors::add))
+                .filter(Objects::nonNull)
+                .collect(Collectors.toList());
+
+        if (modifiedImps.isEmpty()) {
+            return Result.withErrors(errors);
+        }
+
+        final HttpRequest<BidRequest> httpRequest = makeRequest(bidRequest, modifiedImps);
+        return Result.of(List.of(httpRequest), errors);
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/bidder/adf/AdfBidder.java:66]: For one object pls use Collections.singletonList

File: src/main/java/org/prebid/server/bidder/adf/AdfBidder.java [hunk 5]
```diff
@@ -1,56 +1,162 @@
...
+    }
+
+    private Imp modifyImp(Imp imp, Consumer<BidderError> onError) {
+        final Optional<ExtImpAdf> optionalAdfImp;
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/bidder/adf/AdfBidder.java:70]: For now we don't use Optional in our code. Pls update

File: src/main/java/org/prebid/server/bidder/adf/AdfBidder.java [hunk 6]
```diff
@@ -1,56 +1,162 @@
...
+
+        if (modifiedImps.isEmpty()) {
+            return Result.withErrors(errors);
+        }
+
+        final HttpRequest<BidRequest> httpRequest = makeRequest(bidRequest, modifiedImps);
+        return Result.of(List.of(httpRequest), errors);
+    }
+
+    private Imp modifyImp(Imp imp, Consumer<BidderError> onError) {
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/bidder/adf/AdfBidder.java:69]: According to style, pls just pass errors. And then call errors.add

File: src/main/java/org/prebid/server/bidder/adf/AdfBidder.java [hunk 7]
```diff
@@ -1,56 +1,162 @@
...
+            return Result.withError(BidderError.badServerResponse(e.getMessage()));
         }
-
-        throw new PreBidException(String.format("Failed to parse bid %s mediatype", impId));
     }
 
-    private BidType getExtBidPrebidType(ObjectNode bidExt) {
-        final ExtBidPrebid extBidPrebid = getExtBidPrebid(bidExt);
-        return extBidPrebid != null ? extBidPrebid.getType() : null;
+    private BidResponse parseBidResponse(HttpCall<BidRequest> httpCall) throws DecodeException {
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/bidder/adf/AdfBidder.java:113]: According to other bidders style, pls move it to makeBidds method

File: src/main/java/org/prebid/server/bidder/adf/AdfBidder.java [hunk 8]
```diff
@@ -1,56 +1,162 @@
...
+        return bidResponse.getSeatbid().stream()
+                .filter(Objects::nonNull)
+                .map(SeatBid::getBid)
+                .filter(Objects::nonNull)
+                .flatMap(Collection::stream)
+                .filter(Objects::nonNull)
+                .map(bid -> makeBidderBid(bid, bidResponse.getCur(), onError))
+                .filter(Objects::nonNull)
+                .collect(Collectors.toList());
+    }
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/bidder/adf/AdfBidder.java:136]: Pls check other Bidders code style, like between bidder, and update your code style to be as much similar as possible

**Review by snahornyi — CHANGES_REQUESTED** [review]

**Review by snahornyi — APPROVED** [review]

**Review by rpanchyk — APPROVED** [review]

**Pull request merged**